### PR TITLE
Idt feature lc 1305 bookings remove xapi

### PR DESCRIPTION
**Pull request by jaylco**
- New endpoint **PATCH /course_records**
    - Implements JSONPatch (https://en.wikipedia.org/wiki/JSON_Patch) to service partial updates to the course record table. Heavily based on https://cassiomolin.com/2019/06/10/using-http-patch-in-spring/
    - Added a course record input class which limits the available fields for patching
        - currently limited to `state`, `preference` and `required`
- New endpoint **PATCH /modules_records**
    - Input class which allows `state`, `result`, `completionDate`, `score`, `rated`, `paymentMethod`, `paymentDetails` and `bookingStatus` to be updated
- New endpoint **POST /course_records**
    - Facilitates the creation of course records, optionally allows module records to be created as well
- New endpoint **POST /module_records**
    - Facilitates the creation of module records
- New endpoint **GET /course_records**
    - Facilitates getting course record functionality, optionally returning module records
- New course record controller class
- New course record service class
- New Enum validator (https://www.baeldung.com/javax-validations-enums)
- Added a new generic response class for failures, includes a `code` field which can be used at a later date to populate learner-record-specific error codes
- Added new Enum for Course Record Preference
- **Upgraded Gradle to 4.6**
    - This was due to the inclusion of **MapStruct**, the annotations will not work correctly without Gradle 4.6
    - Gradle 4.6 also requires Lombok `annotationProcessor` to be defined 
- Removal of anything Learning Locker related
    - Sync - This is no longer necessary; the frontend now makes decisions on how to update the API in realtime
    - Actions - these now reside in the frontend to support the new business logic
    - Learner record refresh - not required due to point 1
    - MongoDB - the function of this was to delete old data according to GDPR; TTL indexes can be used within Cosmos to achieve this

Please see JIRA notes for more information

### CODE REVIEW

File: src/main/java/uk/gov/cslearning/record/api/input/POST/PostModuleRecordInput.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    @NotBlank(message = "moduleId is required")
+    private String moduleId;
+
+    @NotBlank(message = "ModuleTitle is required")
+    private String moduleTitle;
+
+    private BigDecimal cost;
+
+    @NotNull(message = "optional is required")
+    private Boolean optional;
```
Comment (SavvasCO) [inline src/main/java/uk/gov/cslearning/record/api/input/POST/PostModuleRecordInput.java:36]: `isOptional` seems like a more suitable name.

File: src/main/java/uk/gov/cslearning/record/api/mapper/CourseRecordMapper.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        LocalDateTime now = LocalDateTime.now();
+        CourseRecord cr = new CourseRecord(inputCourse.getCourseId(), inputCourse.getUserId());
+        cr.setLastUpdated(now);
+        inputCourse.getModuleRecords().forEach(mr -> {
+            ModuleRecord convertedModuleRecord = moduleRecordMapper.postInputAsModule(mr);
+            convertedModuleRecord.setCreatedAt(now);
+            convertedModuleRecord.setUpdatedAt(now);
+            cr.addModuleRecord(convertedModuleRecord);
+        });
+        udpateFromPost(cr, inputCourse);
```
Comment (SavvasCO) [inline src/main/java/uk/gov/cslearning/record/api/mapper/CourseRecordMapper.java:42]: Typo

File: src/main/java/uk/gov/cslearning/record/repository/CourseRecordRepository.java [hunk 3]
```diff
@@ -31,6 +31,12 @@ public interface CourseRecordRepository extends JpaRepository<CourseRecord, Long
     @Query("SELECT cr FROM CourseRecord cr JOIN cr.moduleRecords mr where mr.eventId is not null")
     List<CourseRecord> listEventRecords();
 
+    @Query("SELECT cr FROM CourseRecord cr WHERE cr.identity.userId = ?1 AND cr.identity.courseId = ?2")
+    Optional<CourseRecord> getCourseRecord(String userId, String courseId);
+
+    @Query("SELECT cr FROM CourseRecord cr WHERE cr.identity.userId = ?1 AND (cr.identity.courseId = ?2 OR ?2 is NULL)")
+    List<CourseRecord> findByUserIdAndCourseId(String userId, String courseId);
```
Comment (SavvasCO) [inline src/main/java/uk/gov/cslearning/record/repository/CourseRecordRepository.java:38]: Unsure how these 2 methods are different.

File: src/main/java/uk/gov/cslearning/record/service/UserRecordService.java [hunk 4]
```diff
@@ -29,77 +25,37 @@
...
     @Transactional
     public void deleteUserRecords(String uid) {
-        collectionsService.deleteAllByLearnerUid(uid);
         courseRecordRepository.deleteAllByUid(uid);
     }
 
     @Transactional
-    public void deleteOldRecords(DateTime dateTime, LocalDateTime localDateTime) {
-        collectionsService.deleteAllByAge(dateTime);
+    public void deleteOldRecords(LocalDateTime localDateTime) {
```
Comment (SavvasCO) [inline src/main/java/uk/gov/cslearning/record/service/UserRecordService.java:58]: Naming might cause confusion when calling the method. Maybe something like `deleteRecordsLastUpdatedBefore(localDateTime)` would be more suitable?

File: src/test/java/uk/gov/cslearning/record/api/CourseRecordControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test
+    public void testInvalidContentType() throws Exception {
+        String inputJson = "[{ \"op\": \"replace\", \"path\": \"/state\", \"value\": \"COMPLETED\" }\n]";
+        MockHttpServletRequestBuilder builtPatch = buildPatch(inputJson);
+        builtPatch.contentType("application/json");
+        mockMvc.perform(builtPatch)
+                .andExpect(status().isUnsupportedMediaType());
+    }
```
Comment (SavvasCO) [inline src/test/java/uk/gov/cslearning/record/api/CourseRecordControllerTest.java:84]: Would this also need a test to check if a course record is patched successfully?

File: src/main/java/uk/gov/cslearning/record/api/input/POST/PostModuleRecordInput.java (hunk 1 above)
Comment (jaylco) [reply src/main/java/uk/gov/cslearning/record/api/input/POST/PostModuleRecordInput.java:36]: 'optional' reflects the name used in the database which is intentional

Comment (SavvasCO) [reply src/main/java/uk/gov/cslearning/record/api/input/POST/PostModuleRecordInput.java:36]: Ah ok. Fair enough

File: src/main/java/uk/gov/cslearning/record/repository/CourseRecordRepository.java (hunk 3 above)
Comment (SavvasCO) [reply src/main/java/uk/gov/cslearning/record/repository/CourseRecordRepository.java:38]: Ah ok, cool. 🙂

**Review by SavvasCO — APPROVED** [review]

**Pull request merged**